src/topology/metric_space/emetric_space.lean

```diff
@@ -644,6 +652,136 @@ theorem totally_bounded_iff' {s : set α} :
+    calc 2⁻¹ ^ n = 1 * 2⁻¹ ^ n : (one_mul _).symm
+    ... ≤ 3 * 2⁻¹ ^ n : ennreal.mul_le_mul _ le_rfl,
+    -- TODO: use `norm_num`
+    have : ((1 : ℕ) : ℝ≥0∞) ≤ (3 : ℕ), from ennreal.coe_nat_le_coe_nat.2 (by norm_num1),
+    exact_mod_cast this },
+  -- Let us show the rest of the properties. Since the definition expects a family indexed
+  -- by a single parameter, we use `ℕ × ι` as the domain.
+  refine ⟨ℕ × ι, λ ni, D ni.1 ni.2, λ _, Dopen _ _, _, _, λ ni, ⟨ni.2, HDS _ _⟩⟩,
+  -- The sets `D n i` cover the whole space as we proved earlier
+  { refine Union_eq_univ_iff.2 (λ x, _),
+    rcases Dcov x with ⟨n, i, h⟩,
+    exact ⟨⟨n, i⟩, h⟩ },
+  { /- Let us prove that the covering `D n i` is locally finite. Take a point `x` and choose
+    `n`, `i` so that `x ∈ D n i`. Since `D n i` is an open set, we can choose `k` so that
+    `B = ball x (1 / 2 ^ (n + k + 1)) ⊆ D n i`. -/
+    intro x,
+    rcases Dcov x with ⟨n, i, hn⟩,
+    have : D n i ∈ 𝓝 x, from mem_nhds_sets (Dopen _ _) hn,
+    rcases (nhds_basis_uniformity uniformity_basis_edist_inv_two_pow).mem_iff.1 this
+      with ⟨k, -, hsub : ball x (2⁻¹ ^ k) ⊆ D n i⟩,
+    set B := ball x (2⁻¹ ^ (n + k + 1)),
+    refine ⟨B, ball_mem_nhds _ (pow_pos _), _⟩,
+    -- The sets `D m i`, `m > n + k`, are disjoint with `B`
+    have Hgt : ∀ (m ≥ n + k + 1) (i : ι), disjoint (D m i) B,
+    { rintros m hm i y ⟨hym, hyx⟩,
+      rcases memD.1 hym with ⟨z, rfl, hzi, H, hz⟩,
+      have : z ∉ ball x (2⁻¹ ^ k), from λ hz, H n (by linarith) i (hsub hz), apply this,
+      calc edist z x ≤ edist y z + edist y x : edist_triangle_left _ _ _
+      ... < (2⁻¹ ^ m) + (2⁻¹ ^ (n + k + 1)) : ennreal.add_lt_add hz hyx
+      ... ≤ (2⁻¹ ^ (k + 1)) + (2⁻¹ ^ (k + 1)) :
+        add_le_add (hpow_le $ by linarith) (hpow_le $ by linarith)
+      ... = (2⁻¹ ^ k) : by rw [← two_mul, h2pow] },
+    -- For each `m ≤ n + k` there is at most one `j` such that `D m j ∩ B` is nonempty.
+    have Hle : ∀ m ≤ n + k, set.subsingleton {j | (D m j ∩ B).nonempty},
+    { rintros m hm j₁ ⟨y, hyD, hyB⟩ j₂ ⟨z, hzD, hzB⟩,
+      by_contra h,
+      wlog h : j₁ < j₂ := ne.lt_or_lt h using [j₁ j₂ y z, j₂ j₁ z y],
+      rcases memD.1 hyD with ⟨y', rfl, hsuby, -, hdisty⟩,
+      rcases memD.1 hzD with ⟨z', rfl, -, -, hdistz⟩,
+      suffices : edist z' y' < 3 * 2⁻¹ ^ m, from nmem_of_lt_ind h (hsuby this),
+      calc edist z' y' ≤ edist z' x + edist x y' : edist_triangle _ _ _
+      ... ≤ (edist z z' + edist z x) + (edist y x + edist y y') :
+        add_le_add (edist_triangle_left _ _ _) (edist_triangle_left _ _ _)
+      ... < (2⁻¹ ^ m + 2⁻¹ ^ (n + k + 1)) + (2⁻¹ ^ (n + k + 1) + 2⁻¹ ^ m) :
+        by apply_rules [ennreal.add_lt_add]
+      ... = 2 * (2⁻¹ ^ m + 2⁻¹ ^ (n + k + 1)) : by simp only [two_mul, add_comm]
+      ... ≤ 2 * (2⁻¹ ^ m + 2⁻¹ ^ (m + 1)) :
+        ennreal.mul_le_mul le_rfl $ add_le_add le_rfl $ hpow_le (add_le_add hm le_rfl)
+      ... = 3 * 2⁻¹ ^ m : _,
+      rw [mul_add, h2pow, bit1, add_mul, one_mul] },
```

Comment:
What's the point in deferring the proof obligation of the last line of a calc block?

Reply:
Fixed